### ISSUE TITLE
Experiment: Serverless GraphQL, with Terraform & CircleCI

### ISSUE THREAD

**Issue by DFurnes** [DoSomething/infrastructure#129]
We use GraphQL to power our submission & campaign galleries, and our new user profiles. Starting [this quarter](https://docs.google.com/document/d/1Z0LesRih2rql2mumuM3126kkZC2Qw4IbCHMAl7iw77A/edit), we're also using GraphQL to fetch data for broadcasts in Gambit. As we send more traffic to GraphQL, we want to make sure things keep running smoothly!

One of the difficult things about scaling DoSomething.org is that our traffic is not consistent. For example, here's what traffic to [`graphql.dosomething.org`](https://graphql.dosomething.org) has looked like over the past 3 days:

![production-rpm-last-3-days](https://user-images.githubusercontent.com/583202/52079553-d827fc80-2563-11e9-913b-bfb898186d65.png)

Rather than have systems over-provisioned during low-traffic periods (or require manual intervention to scale up and down), I wanted to explore using "serverless" techologies – Lambda and DynamoDB – to scale up and down transparently and only pay for what we use.

I also wanted to take the opportunity to see whether we could manage our serverless applications with Terraform, since that's where we've begun collecting the rest of our infrastructure. Our previous foray into this world, [Bertly](https://github.com/DoSomething/bertly), was built with the [Serverless framework](https://serverless.com) which provides a batteries-included starting point but locks us into provisioning via per-app CloudFormation templates.

There are a few downsides to this forced dependency on CloudFormation:
1. CloudFormation is only able to provision AWS resources, so if we wanted to depend on a resource not built by Amazon, it'd have to be provisioned and hooked up by hand.
2. We lose the ability to connect to other Terraform resources automatically (via module inputs and outputs), and therefore increase the chance of accidentally connecting the "wrong wires" together.
3. Because Serverless provisions app infrastructure as part of the deploy process, it means any developer (or CI service) that's able to deploy a serverless application would also have read & write access to a large portion of our AWS resources. Yikes!
4. It adds conceptual overhead – instead of having everything in one place, it's one more tool to learn.

## Goal
With that in mind, I set out to move our [GraphQL service](https://graphql.dosomething.org) to run on "vanilla" Lambda, using Terraform to provision the deployment bucket, function, roles, logging, and API gateway.

This work was mostly tracked in #119, DoSomething/graphql#46, and #128. 

## Process
Here's a high-level overview of the work that went into this:

🛋 I configured [a re-usable `lambda_function` module](https://github.com/DoSomething/infrastructure/pull/119#issuecomment-457675183) that provisions a deploy bucket (seeded with a "hello world" example app), API Gateway w/ Lambda proxy, Lambda function, CloudWatch log buckets, a base execution role, and a minimally priviliged access key for deploys.

🥞 I added support for [swappable cache drivers](https://github.com/DoSomething/graphql/pull/48), so we could use Redis (on Heroku) and DynamoDB (on Lambda) with the same backend code. DynamoDB offers on-demand scaling and pricing (and [no VPC requirement](https://github.com/DoSomething/infrastructure/pull/119#issuecomment-457675183)), which fits well with the serverless philosophy we're going for here.

🕵️ I ran into some spooky timeouts & dug into [why they might be happening with X-Ray](https://github.com/DoSomething/infrastructure/pull/119#issuecomment-457762658). It turned out there were two culprits: occasionally [slow cold-boots when using less RAM](https://github.com/DoSomething/infrastructure/pull/125) and [an issue with our module loader](https://github.com/DoSomething/graphql/pull/46#issuecomment-458720632).

📦 Using the Terraform module, I easily [spun up QA & production environments](https://github.com/DoSomething/infrastructure/pull/128) for GraphQL.

🤹‍♂️ I configured [CircleCI](https://circleci.com) to [run our builds & deploy to each Lambda environment](https://github.com/DoSomething/graphql/pull/46#issuecomment-459106455) (using the minimally priviliged AWS access key provisioned by Terraform). Plus, [a handy approval UI for prod deploys](https://github.com/DoSomething/graphql/pull/52#issuecomment-460441073)!

## Benchmarks
Since one of the goals here was scalability & performance, I did a lot of load testing with [Siege](https://www.joedog.org/siege-manual/).

I've summarized the results below, comparing a `topic` query (used by Gambit & hitting a local cache), and a `postsByCampaignId` query (used by Phoenix's campaign gallery & hitting Rogue and Northstar).

I ran these tests from `mjolnir.dosomething.org` against our QA environment, with applicable Heroku dynos temporarily scaled up to match their production capacity. To make sure the results were representative, I ran each test until it completed 50,000 requests.

#### Broadcast query (simple)
|                           | Duration | Throughput | Average | Worst | Comparison                          |
|--------------------------------|----------|------------|---------|-------|:-----------------------------:|
| **Lambda backend** &nbsp;&nbsp;&nbsp;&nbsp;  | 1.6m     | 518.4 rps  | 0.08s   | 7.71s | :trophy:          |                  
| **Heroku backend**             | 3.2m     | 257.9 rps  | 0.18s   | 1.27s | :small_red_triangle_down: 125% |

<details>
<summary>raw <code>topic</code> test results</summary>

```
siege -c50 -r1000 --content-type "application/json" 'https://rpkpk1dw7b.execute-api.us-east-1.amazonaws.com/qa POST {"query":"{topic(id: \"G8LCAYO9WuGC2UW8mQi6w\") { name,  ...on AskYesNoBroadcastTopic {text,saidYes}}}"}'

Transactions:		       50000 hits
Availability:		      100.00 %
Elapsed time:		       96.45 secs
Data transferred:	       39.34 MB
Response time:		        0.08 secs
Transaction rate:	      518.40 trans/sec
Throughput:		        0.41 MB/sec
Concurrency:		       42.51
Successful transactions:       50000
Failed transactions:	           0
Longest transaction:	        7.71
Shortest transaction:	        0.02



siege -c50 -r1000 --content-type "application/json" 'https://graphql-qa.dosomething.org/graphql POST {"query":"{topic(id: \"G8LCAYO9WuGC2UW8mQi6w\") { name,  ...on AskYesNoBroadcastTopic {text,saidYes}}}"}'

Transactions:		       49999 hits
Availability:		      100.00 %
Elapsed time:		      193.82 secs
Data transferred:	       39.34 MB
Response tme:		        0.18 secs
Transaction rate:	      257.97 trans/sec
Throughput:		        0.20 MB/sec
Concurrency:		       46.32
Successful transactions:       49999
Failed transactions:	           1
Longest transaction:	        1.27
Shortest transaction:	        0.02
```

</details>

#### Campaign Gallery query (complex)
|                           | Duration | Throughput | Average | Worst | Comparison                          |
|---------------------------------|----------|------------|---------|-------|:-----------------------------:|
| **Lambda backend** &nbsp;&nbsp;&nbsp;&nbsp; | 21.6m    | 38.6 rps   | 1.28s   | 5.90s |    :trophy:      |                
| **Heroku backend** | 40.4m    | 20.6 rps   | 2.41s   | 9.48s | :small_red_triangle_down: 88% |

<details>
<summary>raw <code>postsByCampaignId</code> test results</summary>

```
siege -c50 -r1000 --content-type "application/json" 'https://rpkpk1dw7b.execute-api.us-east-1.amazonaws.com/qa POST {"query":"query CampaignGalleryQuery($campaignId: String!, $count: Int, $page: Int) {\npostsByCampaignId(id: $campaignId, count: $count, page: $page) {...PostCard,...ReactionButton,__typename}}\nfragment PostCard on Post {id,type,status,url,text,tags,quantity,user {firstName,__typename},__typename}\nfragment ReactionButton on Post {reactions, reacted,  __typename}","variables":{"campaignId":"5646","count":12,"page":1}}'

Transactions:		       50000 hits
Availability:		      100.00 %
Elapsed time:		     1295.03 secs
Data transferred:	      159.26 MB
Response time:		        1.28 secs
Transaction rate:	       38.61 trans/sec
Throughput:		        0.12 MB/sec
Concurrency:		       49.39
Successful transactions:       50000
Failed transactions:	           0
Longest transaction:	        5.90
Shortest transaction:	        0.18



siege -c50 -r1000 --content-type "application/json" 'https://graphql-qa.dosomething.org/graphql POST {"query":"query CampaignGalleryQuery($campaignId: String!, $count: Int, $page: Int) {\npostsByCampaignId(id: $campaignId, count: $count, page: $page) {...PostCard,...ReactionButton,__typename}}\nfragment PostCard on Post {id,type,status,url,text,tags,quantity,user {firstName,__typename},__typename}\nfragment ReactionButton on Post {reactions, reacted,  __typename}","variables":{"campaignId":"5646","count":12,"page":1}}'

Transactions:		       49999 hits
Availability:		      100.00 %
Elapsed time:		     2423.95 secs
Data transferred:	      159.26 MB
Response time:		        2.41 secs
Transaction rate:	       20.63 trans/sec
Throughput:		        0.07 MB/sec
Concurrency:		       49.77
Successful transactions:       49999
Failed transactions:	           1
Longest transaction:	        9.48
Shortest transaction:	        0.15
```

</details>
<br>

In both cases, the Lambda function is able to handle significantly more load than our Heroku dyno, without any human intervention to scale up resources. The gallery query hits a performance boundary based on Rogue & Northstar's capacity, but still performs nearly twice as well.

I also looked at the latency distribution for these queries (measured with Apollo Engine), which is harder to compare side-by-side but still somewhat interesting: [`topic` (Heroku)](https://user-images.githubusercontent.com/583202/52143409-5f8b7380-2629-11e9-808e-92379479ce5c.png), [`topic` (Lambda)](https://user-images.githubusercontent.com/583202/52143418-65815480-2629-11e9-9c97-3eca9c33f03e.png), [`postsByCampaignId` (Heroku)](https://user-images.githubusercontent.com/583202/52143394-53071b00-2629-11e9-8d67-9c321111abc5.png), [`postsByCampaignId` (Lambda)](https://user-images.githubusercontent.com/583202/52143428-6d40f900-2629-11e9-9c1b-a927358d1983.png). 🤓 

## Conclusion
I'm really happy with how this experiment has turned out so far! In summary:

**Performance:** There were some massive performance wins. For both light and heavy loads, we roughly doubled the amount of throughput we could handle with this application. Alongside effortless horizontal scaling, we also get the peace-of-mind of pay-per-request pricing, so we don't end up paying for beefy servers we don't use during lulls.

**Workflow:** I think the workflow for provisioning new serverless apps with Terraform & then deploying them with CI works really smoothly, and has some nice ergonomic & security benefits over how the Serverless framework does things: no credentials on local machines, deploy credentials are limited to _just_ deploying, and consolidating our infrastructure into one tool.

**Bonus:** I tried out [CircleCI](https://circleci.com) as a continuous integration tool here (instead of Wercker or Jenkins). I'll make a separate issue to track progress on that, but the **tl;dr** is that I'd love to get us on a more industry-standard CI service & be able to move more of our jobs (including scheduled jobs, like DB refreshes) into code. None of the work done in this experiment is _inherently_ tied to CircleCI, though.

With this in mind, I think it's worth thinking about where else we could use Serverless to simplify our infrastructure. Two places that come to mind are Blink (which is the poster child for uneven traffic spikes!) and Rogue's image formatting endpoint (which has unusually CPU and memory-intensive, but infrequent, work each time it needs to render an uncached image).

**Comment by DFurnes** [comment]
Finally, a checklist of some things that still need investigation before we can consider this "shipped": 

- [ ] Are there changes we'll need to make our Terraform template more flexible for other apps? What about Python? Applications that don't use the generic "proxy" route?
- [ ] How will custom domains work, especially if we don't have the DNS record purchased in Route 53?
- [ ] How should our GraphQL playground & authentication flow work in a serverless world?
- [ ] We need to hook up these log groups to our [Papertrail log forwarder](https://github.com/DoSomething/lambda-papertrail).

**Comment by DFurnes** [comment]
I've made some updates to our CircleCI config in DoSomething/graphql#52. The deployment graph is now [a little clearer](https://github.com/DoSomething/graphql/pull/52#issuecomment-460308794), CircleCI automatically loads deploy credentials via Amazon SSM, and deployment logic has been extracted to the [`dosomething/lambda`](https://github.com/DoSomething/lambda-orb) so it can be shared between projects.

**Comment by mshmsh5000** [comment]
Probably good to point out _orbs_, which are a CircleCI-specific entity—a modular workflow package:

https://circleci.com/docs/2.0/orb-intro/

>CircleCI Orbs are shareable packages of configuration elements, including jobs, commands, and executors. CircleCI provides certified orbs, along with 3rd-party orbs authored by CircleCI partners. It is best practice to first evaluate whether any of these existing orbs will help you in your configuration workflow.

**Comment by DFurnes** [comment]
> How will custom domains work, especially if we don't have the DNS record purchased in Route 53?

It looks like the answer is "easily"! We used a [manually provisioned](https://dosomething.slack.com/archives/C03T8SDJJ/p1549405167032700) ACM wildcard certificate (since it requires us to attach a CNAME to our domain for validation) and are now able to set a custom domain for each app via #133. Each CNAME target is available on the [Custom Domain Names](https://console.aws.amazon.com/apigateway/home?region=us-east-1#/custom-domain-names) page.

**Comment by DFurnes** [comment]
> We need to hook up these log groups to our [Papertrail log forwarder](https://github.com/DoSomething/lambda-papertrail).

This is done in #134, and we can now easily forward any other function's logs using the existing forwarder function and the new (optional) `logger` parameter on the `lambda_function` module.

**Comment by DFurnes** [comment]
> The `lambda_function` module needs docs, and ideally a high-level "getting started" guide.

I added a "getting started" guide for this (and demo'd at Sprint Retro) in #137, and wrote up more detailed usage instructions for the `lambda_function` and `api_gateway_proxy` modules in #143.

**Comment by DFurnes** [comment]
> Are there changes we'll need to make our Terraform template more flexible for other apps? What about Python?

Added support for `python2.7` and `python3.7` runtimes in #143 as well.

**Comment by DFurnes** [comment]
> Applications that don't use the generic "proxy" route?

Finally, tackled this one in #146 by creating a generic `api_gateway` module which can accepts a list of function & path mappings (so a single "app" could be composed from a bunch of lambdas).

**Comment by DFurnes** [comment]
Fixed up some tiny bugeroos in https://github.com/DoSomething/graphql/commit/b6f2db1b93ab9f7b91b44bba1bfcef2a1f98ec4f, https://github.com/DoSomething/graphql/pull/58, and https://github.com/DoSomething/infrastructure/pull/148. I think we're ready to swap these in place of the Heroku backends on dev/QA & see how it goes! :v:

**Comment by DFurnes** [comment]
Deleted the old Heroku/Fastly resources for this application in #157, since things have been happily running on Lambda for around a week now. If we need to spin this back up in the future, it should be as easy as reverting that pull request.

**Comment by DFurnes** [comment]
> How should our GraphQL playground & authentication flow work in a serverless world? 

I've taken a stab at this in DoSomething/northstar#840.

>  How should offline development work for these applications? What about when multiple functions are composed into a single app via API Gateway?

I'd looked into [serverless-offline](https://github.com/dherault/serverless-offline), which unfortunately requires a Serverless framework template, and [localstack](https://github.com/localstack/localstack) which seemed promising, but again didn't have a clear way to "wire things up".

In the meantime, I made some cleanups for consistency between local dev and production in DoSomething/graphql#62 so this is more consistent for developers.

**Comment by DFurnes** [comment]
Since this experiment is effectively wrapped up, I'm going to close this issue out!
